iwyu.cc

```diff
@@ -1375,62 +1375,55 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
   SourceLocation GetCanonicalUseLocation(SourceLocation use_loc,
                                          const NamedDecl* decl) {
     // If we're not in a macro, just echo the use location.
-    if (!IsInMacro(use_loc))
+    if (!use_loc.isMacroID())
       return use_loc;
 
-    // If the macro definition file contains a forward-declaration for decl,
-    // we treat that as a use-attribution hint. See below for details.
-    const FileEntry* macro_def_file = GetFileEntry(GetSpellingLoc(use_loc));
+    VERRS(5) << "Trying to determine use location for '"
+             << PrintableDecl(decl) << "'\n";
+
+    clang::SourceManager* sm = GlobalSourceManager();
+    SourceLocation spelling_loc = sm->getSpellingLoc(use_loc);
+    SourceLocation expansion_loc = sm->getExpansionLoc(use_loc);
+
+    // If the file defining the macro contains a forward decl, keep it around
+    // and treat it as a hint that the expansion loc is responsible for the
+    // symbol.
+    const FileEntry* macro_def_file = GetLocFileEntry(spelling_loc);
+    VERRS(5) << "Macro is defined in file '" << GetFilePath(macro_def_file)
+             << "'. Looking for fwd-decl hint...\n";
+
     const NamedDecl* fwd_decl = nullptr;
     for (const NamedDecl* redecl : GetClassRedecls(decl)) {
-      if (GetFileEntry(redecl) == macro_def_file &&
-          IsForwardDecl(redecl)) {
+      if (GetFileEntry(redecl) == macro_def_file && IsForwardDecl(redecl)) {
         fwd_decl = redecl;
+
+        // Make sure we keep that forward-declaration, even if it's probably
+        // unused in this file.
+        IwyuFileInfo* file_info =
+            preprocessor_info().FileInfoFor(macro_def_file);
+        file_info->ReportForwardDeclareUse(
+            spelling_loc, fwd_decl,
+            IsNodeInsideCXXMethodBody(current_ast_node()), nullptr);
```

Comment:
I suspect this `ReportForwardDeclareUse` isn't covered by tests because I tried with `use_loc` and `decl` and nothing was broken. It's not a blocker, just mention it for general awareness.